src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+
+                    int prevConsumed = _consumed;
+                    long prevPosition = _lineBytePosition;
+                    long prevLineNumber = _lineNumber;
+                    retVal = ConsumePropertyName();
+                    if (!retVal)
+                    {
+                        // roll back potential changes
+                        _consumed = prevConsumed;
+                        _tokenType = JsonTokenType.StartObject;
+                        _lineBytePosition = prevPosition;
+                        _lineNumber = prevLineNumber;
+                    }
+                    goto Done;
+                }
+            }
+            else if (_tokenType == JsonTokenType.StartArray)
+            {
+                if (first == JsonConstants.CloseBracket)
+                    EndArray();
+                else
+                {
+                    retVal = ConsumeValue(first);
+                    goto Done;
+                }
+            }
+            else if (_tokenType == JsonTokenType.PropertyName)
+            {
+                retVal = ConsumeValue(first);
+                goto Done;
+            }
+            else
+            {
+                int prevConsumed = _consumed;
+                long prevPosition = _lineBytePosition;
+                long prevLineNumber = _lineNumber;
+                JsonTokenType prevTokenType = _tokenType;
+                ConsumeTokenResult result = ConsumeNextToken(first);
+                if (result == ConsumeTokenResult.Success)
+                {
+                    retVal = true;
+                    goto Done;
+                }
+                if (result == ConsumeTokenResult.IncompleteRollback)
+                {
+                    _consumed = prevConsumed;
+                    _tokenType = prevTokenType;
+                    _lineBytePosition = prevPosition;
+                    _lineNumber = prevLineNumber;
+                }
+                goto Done;
+            }
+
+            retVal = true;
+
+        Done:
+            return retVal;
+
+        ReadFirstToken:
+            retVal = ReadFirstToken(first);
+            goto Done;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool HasMoreData()
+        {
+            if (_consumed >= (uint)_buffer.Length)
+            {
+                if (_isNotPrimitive && IsLastSpan)
+                {
+                    if (_tokenType != JsonTokenType.EndArray && _tokenType != JsonTokenType.EndObject)
+                        ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.InvalidEndOfJson);
+                }
+                return false;
+            }
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool HasMoreData(ExceptionResource resource)
+        {
+            if (_consumed >= (uint)_buffer.Length)
+            {
+                if (IsLastSpan)
```

Comment:
The shape of the body of this overload is different than the other one.  I can't say if either seems more right, but "different" seems like "possibly wrong" or "add some comments to explain the difference"